EIPS/eip-4519.md

```diff
@@ -0,0 +1,239 @@
+    /// @return the owner of the token bound to the asset that generates _addressSA.
+    function ownerOfFromBCA(address _addressSA) external view returns (address);
+    
+    /// @notice This function lets know the user of the token from its tokenID.
+    /// @dev Everybody can call this function. The code executed only reads from the blockchain.
+    /// @param _tokenId is the tokenID of the SmartNFT bound to the asset.
+    /// @return The user of the token from its _tokenId.
+    function userOf(uint256 _tokenId) external view returns (address);
+    
+    /// @notice This function lets know the user of the token from the address of the asset bound to the token.
+    /// @dev Everybody can call this function. The code executed only reads from the blockchain.
+    /// @param _addressSA is the address to obtain the user from it.
+    /// @return The user of the token bound to the asset that generates _addressSA.
+    function userOfFromBCA(address _addressSA) external view returns (address);
+    
+    /// @notice This function lets know how many tokens are assigned to a user.
+    /// @dev Everybody can call this function. The code executed only reads from the blockchain.
+    /// @param _addressUser is the address of the user.
+    /// @return the number of tokens assigned to a user.
+    function userBalanceOf(address _addressUser) external view returns (uint256);
+    
+    /// @notice This function lets know how many tokens of an particular owner are assigned to a user.
+    /// @dev Everybody can call this function. The code executed only reads from the blockchain.
+    /// @param _addressUser is the address of the user.
+    /// @param _addressOwner is the address of the owner.
+    /// @return the number of tokens assigned to a user from an owner.
+    function userBalanceOfAnOwner(address _addressUser, address _addressOwner) external view returns (uint256);
+}
+```
+This interface is an extension of the ERC-721, is compatible with the standard, and needs the ERC-721 interface. ERC-721 interface is available in https://eips.ethereum.org/EIPS/eip-721. Then, metadata and enumeration extensions are compatible and included in this draft.
+Like EIP-165 is required, supportsInterface SHOULD BE overridden to include the interfaceId of this EIP.
+ 
+## Rationale
+This SmartNFT was developed because the ERC-721 standard does not define the "users" of an asset, only the "owners", and does not assign a blockchain account (BCA) address, "asset", to the physical asset. Smart assets (for example IoT devices) are increasing nowadays and they can be managed in a secure way if they are bound to SmartNFTs. SmartNFTs allow implementing a secure way to share a secret key between the owner and the asset and between the user and the asset, confirmed with the consensus of the blockchain. In this way, assets, owners, and users can ensure that they are exchanging information only with trusted parts.
```

Comment:
```suggestion
This SmartNFT was developed because the ERC-721 standard does not define the "users" of an asset, only the "owners", and does not assign a blockchain account (BCA) address, "asset", to the physical asset. Smart assets (for example**,** IoT devices) are increasing nowadays and they can be managed in a secure way if they are bound to SmartNFTs. SmartNFTs allow implementing a secure way to share a secret key between the owner and the asset and between the user and the asset, confirmed with the consensus of the blockchain. In this way, assets, owners, and users can ensure that they are exchanging information only with trusted parts.
```